Add support for SLNX solution file format in NuGet.exe

FXZFun opened the pull request:
<!-- DO NOT MODIFY OR DELETE THIS TEMPLATE. IT IS USED IN AUTOMATION. -->

# Bug

<!-- If this is an engineering change or test change only, you do not need an issue. -->
<!-- Find or create an issue in NuGet/Home and paste the full url. -->
<!-- At the maintainers discretion, multiple changes may apply to a single issue, but only when the PRs are all created within a short period of time. -->
Fixes: https://github.com/nuget/home/issues/14113

## Description
 - Adds support for restoring and updating solutions using the new .SLNX file format. 
 - The only functional changes necessary were modifying the file filtering checks to include .SLNX extensions. 
 - Added tests in the restore and update command tests to verify the changes function properly.

## PR Checklist

- [x] Meaningful title, helpful description and a linked NuGet/Home issue
- [x] Added tests
- [x] ~~Link to an issue or pull request to update docs if this PR changes settings, environment variables, new feature, etc.~~

FXZFun committed:
Refactor solution file checks to add support for .slnx files

- Updated `GetInputsFromFile` to use `IsSolutionFile()` for solution file checks.
- Modified `GetInputsFromDirectory` to utilize `IsSolutionFile()` instead of direct extension checks.
- Streamlined `IsSolutionOrProjectFile` to call `IsSolutionFile()` for identifying solution files.
- Replaced ".sln" extension check in `GetInputFile` with `IsSolutionFile()` for consistency.
- Added `IsSolutionFile` extension method in `ProjectHelper` for better encapsulation of solution file logic.
- Introduced `TryGetProjectFile` method in `ProjectHelper` to find project files in a directory.
- Updated `SolutionFileExists` to use `EnumerateFiles` with `IsSolutionFile()` for flexibility in recognizing solution file types.

FXZFun committed:
Add tests for new .slnx solution file format

- Modify unit tests in `RestoreCommandTests.cs` and `NuGetUpdateCommandTests.cs` to utilize `.slnx` format.
- Update method signatures to accept a boolean parameter for solution type.
- Change file creation methods to generate `.slnx` files instead of `.sln`.
- Adjust project setup to accommodate the new solution file format.
- Add a new test method for updating packages with pre-release versions in the `.slnx` format.
- Introduce a new method in `Util.cs` for creating `.slnx` file content.

FXZFun committed:
Merge remote-tracking branch 'origin/dev' into dev-fxzfun-add-more-slnx-support

FXZFun committed:
Merge branch 'NuGet:dev' into dev-fxzfun-add-more-slnx-support

FXZFun commented:
@dotnet-policy-service agree

FXZFun commented:
@microsoft-github-policy-service agree

nkolev92 commented on test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/RestoreCommandTests.cs:
```diff
@@ -995,14 +995,16 @@ public CommandRunnerResult RunRestore(SimpleTestPathContext pathContext, int exp
+        [Theory]
+        [InlineData(true)]
+        [InlineData(false)]
+        public async Task Restore_PackageSourceMapping_Succeed(bool useSlnx)
```
I think the tests are failing because the 17.12 version of MSBuild is being used. 

17.13 is the first version of MSBuild that supports slnx, is that correct @rainersigwald?

FXZFun replied:
This [devblog](https://devblogs.microsoft.com/dotnet/introducing-slnx-support-dotnet-cli/?commentid=21515#comment-21515) agrees that MSBuild support was introduced in 17.13.
I'm running 17.13.19.12613 on my machine and all the tests passed for me.

nkolev92 replied:
We might be able to raise a more helpful error message if someone tries to use an slnx with an msbuild version older than 17.13. 

Don't feel too strongly right now. 
We do need to either used a different pool for our tests or gate the tests.

nkolev92 replied:
@FXZFun 

I synced with my colleagues and we think it'd be a great idea to recognize when MSBuild is a lower version than 17.13 and error when an slnx is passed. 

As far as the test goes, the CI requirements may prove this test to be unreliable, so maybe it can account for the error or restore depending on the version it can find.

FXZFun committed:
Add validation for .slnx files in ProcessSolutionFile

- Introduced validation for .slnx file extension in ProcessSolutionFile method.
- Checked if MSBuild version is less than 17.13 for .slnx files.
- Threw InvalidOperationException for unsupported MSBuild versions.
- Added localized error message for unsupported .slnx format.
- Updated NuGetResources.Designer.cs with new error string.
- Added corresponding entry in NuGetResources.resx for .slnx error.

nkolev92 reviewed (commented):
Changes look good, but tests are still failing because 17.12 is not available. 

We can probably add a Skip on the slnx legs on the CI so that we can merge this PR.

thoughts @NuGet/nuget-client

FXZFun committed:
Update to skip tests on MSBuild 17.12 and older

- Modified test cases in `RestoreCommandTests.cs`
- Replaced `[InlineData(true)]` with `[InlineData(true, Skip = "MSBuild >= 17.13 not available")]`
- Added skip condition for tests dependent on MSBuild version

FXZFun commented:
Tests are passing, but the dotnet process failed. I haven't had a chance to investigate yet.

FXZFun commented:
All the checks seem to be good here now.

FXZFun commented:
@nkolev92 When can I expect these changes to be implemented?

nkolev92 commented:
@FXZFun It'll be post 6.14, we're a few weeks late for 6.14. 

@NuGet/nuget-client can someone else take a look before merging?
